cmd/snap-update-ns/bootstrap.c

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "bootstrap.h"
+
+#define _GNU_SOURCE
+
+#include <errno.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <sched.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+// bootstrap_errno contains a copy of errno if a system call fails.
+int bootstrap_errno = 0;
+// bootstrap_msg contains a static string if something fails.
+const char* bootstrap_msg = NULL;
+
+// read_cmdline reads /proc/self/cmdline into the specified buffer, returning
+// number of bytes read.
+static ssize_t
+read_cmdline(char* buf, size_t buf_size)
+{
+    int fd = open("/proc/self/cmdline", O_RDONLY | O_CLOEXEC | O_NOFOLLOW);
+    if (fd < 0) {
+        bootstrap_errno = errno;
+        bootstrap_msg = "cannot open /proc/self/cmdline";
+        return -1;
+    }
+    memset(buf, 0, buf_size);
+    ssize_t num_read;
+    if ((num_read = read(fd, buf, buf_size)) < 0) {
+        bootstrap_errno = errno;
+        bootstrap_msg = "cannot read /proc/self/cmdline";
+    }
+    close(fd);
+    return num_read;
+}
+
+// find_snap_name scans the command line buffer and looks for the 1st argument.
+static const char*
+find_snap_name(char* buf, size_t buf_size, size_t num_read)
+{
+    // cmdline is an array of NUL ('\0') separated strings. We can skip over
+    // the first entry (program name) and look at the second entry, in our case
+    // it should be the snap name.
+    size_t argv0_len = strnlen(buf, buf_size);
+    if (argv0_len == num_read) {
+        return NULL;
+    }
+    return &buf[argv0_len + 1];
```

Comment:
If \0 is the last byte, then argv_len < num_read, yet argv0_len+1 is an overrun.